### ISSUE TITLE
[suggestion/request] automatic churn

### ISSUE THREAD

**Issue by aeon1234** [monero-project/monero#4305]
it would be nice to have automatic churn in your wallet, so you can run it 24/7 and it does it for you and uses delays which match good distribution.

it could look like this:

`[4ABCD]: churn`
`starting churn`
`[4ABCD]: transfer 4ABCD... 1`
`ERROR: please stop churn before manually using your wallet again.`

**Comment by jtgrassie** [comment]
Why the need for a churn?

**Comment by SamsungGalaxyPlayer** [comment]
It would be nice at least for the CLI wallet. However, the best process for churning is still debated.

If I had to come up with a primitive technique right now, it would probably include 1) waiting for the output to be included in another transaction as a decoy, 2) waiting a random amount of time between 0.5-3 days, or 3) some combination of the two.

However, I have not tested if this is a good mechanism or not, and I think most mechanisms that people have hypothesized have not been tested yet either.

**Comment by aeon1234** [comment]
is it possible to use same distribution for new generated tx like #3528 does for decoys?

and in general is it not better to have any (semi perfect) automatic churn algorithm instead of people doing it manually?

**Comment by jtgrassie** [comment]
Sorry to harp on about this but can you please explain _why_ there is a need for churning at all (automatic or manual aside).

**Comment by SamsungGalaxyPlayer** [comment]
@jtgrassie I strongly recommend reading #4229, especially [this comment](https://github.com/monero-project/monero/issues/4229#issuecomment-415139034). Churning increases the entropy set more than a single transaction can provide.

**Comment by paulshapiro** [comment]
I'm definitely down to prototype and explore these. MyMonero apps support testnet mode btw.

**Comment by paulshapiro** [comment]
Well, the code does, in any case. Just need to flip it in code

**Comment by paulshapiro** [comment]
Note: deleted my comment b/c I thought this was another repo

**Comment by aeon1234** [comment]
> It certainly needs more thought.

can you clarify what exactly needs more thought?

people are currently manual churning or are maybe using their own scripts (with more or less good algorithm) for this.
so why not implementing it?

**Comment by jtgrassie** [comment]
If you follow the thread sgp posted, it's not clear cut on the best route. Hence more discussion/thought.

**Comment by SamsungGalaxyPlayer** [comment]
@aeon1234 it's because we currently don't have a good idea what the wallet behavior should be. Once we get a better idea, I agree that it should be added.

**Comment by aeon1234** [comment]
on the other side you guys from the MRL have way better understanding what to do than most of us users. also a user cannot create a tx at any time of the day, most have to work and sleep eventually. so most manual churns are probably way more bad than any algorithm from you, even if this algorithm is not perfect for now. so I hope you do not wait until there is a perfect churn behavior in the future.

**Comment by aeon1234** [comment]
there are at least some examples in the linked thread where I understood that churning is needed. or did I just read this wrong and churning is not/never needed?

**Comment by jtgrassie** [comment]
@aeon1234 There is no linked _research_ in that thread regarding churning. You are _assuming_ consensus on churning. If there was solid research there would likely be some consensus, but that is not where we are at now. Most debate has been about ring-size (fixed, size, etc), not churning.

**Comment by jtgrassie** [comment]
> my whole issue assumes that there was already consensus that churning is needed...  if that's wrong...

It's wrong. There is no hard research and no solid consensus. When there is research, there will undoubtedly be some consensus, until then, this kind of suggestion has no merit. If you have done some real research to point to a need for churning or ring-size changes, please add it to your suggestion so it can be debated.

**Comment by SamsungGalaxyPlayer** [comment]
@jtgrassie after being involved in the research meetings, I can confidently state that research points towards churning having a positive or neutral impact in most circumstances, not a negative impact. The only possible negative impact I can think about at the moment is the increased attack surface from potential timing attack heuristics, eg: increasing the number of data points for transactions sent at a certain exact time each day. However, this is exactly something that a wallet feature can mitigate: reducing the feasibility of timing attacks for heuristics we know of. However, as I have stated, these have not been thoroughly defined.

From a strict ring signature perspective, churning increases the number of possible outputs spent in transactions. So, if we can incorporate churning while minimizing the timing and IP address metadata leaked, then it should be a clear win for privacy. However, since we have not defined these, I do not recommend its addition at this time until some heuristics are developed and tested.

In short, research is clearly pointing towards churning having potentially positive effects, but the best way to get these effects is not well-defined. As a result, we shouldn't incorporate this into the wallet yet, since we would need to have these definitions first. However, once we have a better idea, it should clearly be added.

**Comment by selsta** [comment]
There are no plans to implement this feature at the moment, especially with the upcoming FCMP++ ugrade.
